crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -485,18 +497,55 @@ fn zome_id_to_zome(zome_id: ZomeId, dna_def: &DnaDef) -> AppValidationResult<Zom
         .into())
 }
 
-fn run_validation_callback_inner(
+#[async_recursion::async_recursion]
+async fn run_validation_callback_inner<R>(
     invocation: ValidateInvocation,
-    ribosome: &impl RibosomeT,
-    workspace_lock: HostFnWorkspaceRead,
+    ribosome: &R,
+    workspace_read: HostFnWorkspaceRead,
     network: HolochainP2pDna,
-) -> AppValidationResult<Outcome> {
-    let validate: ValidateResult =
-        ribosome.run_validate(ValidateHostAccess::new(workspace_lock, network), invocation)?;
-    match validate {
+    fetched_deps: &mut HashSet<AnyDhtHash>,
+) -> AppValidationResult<Outcome>
+where
+    R: RibosomeT + Sync,
+{
+    let validate_result = ribosome.run_validate(
+        ValidateHostAccess::new(workspace_read.clone(), network.clone()),
+        invocation.clone(),
+    )?;
+    match validate_result {
         ValidateResult::Valid => Ok(Outcome::Accepted),
         ValidateResult::Invalid(reason) => Ok(Outcome::Rejected(reason)),
-        ValidateResult::UnresolvedDependencies(hashes) => Ok(Outcome::AwaitingDeps(hashes)),
+        ValidateResult::UnresolvedDependencies(hashes) => {
+            if !hashes.iter().any(|hash| !fetched_deps.contains(hash)) {
+                Ok(Outcome::AwaitingDeps(hashes))
+            } else {
+                let in_flight = hashes.into_iter().map(|hash| async {
+                    let cascade_workspace = workspace_read.clone();
+                    let mut cascade =
+                        Cascade::from_workspace_network(&cascade_workspace, network.clone());
+                    cascade
+                        .fetch_element(hash.clone(), NetworkGetOptions::must_get_options())
+                        .await?;
+                    Ok(hash)
+                });
+                let results: AppValidationResult<Vec<AnyDhtHash>> =
+                    futures::future::join_all(in_flight)
```

Comment:
This should not be a join_all as they are unbounded. Use a [buffered](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffered) instead and give it a concurrent bound.